CodeEdit/NavigatorSidebar/ProjectNavigator/ProjectNavigatorItem.swift

```diff
@@ -0,0 +1,166 @@
+//
+//  SideBarItem.swift
+//  CodeEdit
+//
+//  Created by Lukas Pistrol on 17.03.22.
+//  Context Menu by Nanashi Li on 20.03.22.
+//
+
+import SwiftUI
+import WorkspaceClient
+import CodeFile
+
+struct ProjectNavigatorItem: View {
+	@Environment(\.controlActiveState) var activeState
+	@Environment(\.colorScheme) var colorScheme
+	@AppStorage(FileIconStyle.storageKey) var iconStyle: FileIconStyle = .default
+
+	/// The `FileItem` for this view
+	var item: WorkspaceClient.FileItem
+
+	/// The current ``WorkspaceDocument``
+	@ObservedObject var workspace: WorkspaceDocument
+
+	/// The current `NSWindowController`
+	var windowController: NSWindowController
+
+	/// If ``item`` is a folder, load one level of children if expanded
+	/// This fixes animation glitches when unfolding a folder while
+	/// maintaining performance.
+	///
+	/// This will be set to true once a parent folder expands
+	@Binding var shouldloadChildren: Bool
+
+	/// The current selected ``item`` in the list
+	@Binding var selectedId: WorkspaceClient.FileItem.ID?
+
+	/// Tracks the `DisclosureGroup`s `isExpanded` property in order
+	/// to set ``shouldloadChildren`` for its child items
+	@State private var isExpanded: Bool = false
+
+	var body: some View {
+		if item.children == nil {
+			sidebarFileItem(item)
+				.id(item)
+		} else {
+			sidebarFolderItem(item)
+				.id(item.id)
+		}
+	}
+
+	func sidebarFileItem(_ item: WorkspaceClient.FileItem) -> some View {
+		HStack {
+			Image(systemName: item.systemImage)
+				.resizable()
+				.aspectRatio(contentMode: .fit)
+				.frame(width: 12, height: 12, alignment: .center)
+				.foregroundColor(iconColor)
+			Text(item.url.lastPathComponent)
+				.font(.subheadline)
+				.frame(maxWidth: .infinity, alignment: .leading)
+				.contentShape(Rectangle())
+		}
+		.listRowInsets(.init())
+		.contextMenu { contextMenuContent(false) }
+	}
+
+	@ViewBuilder
+	func sidebarFolderItem(_ item: WorkspaceClient.FileItem) -> some View {
+		DisclosureGroup(isExpanded: $isExpanded) {
+			if shouldloadChildren { // Only load when expanded -> Improves performance massively
+				ForEach(item.children!.sortItems(foldersOnTop: workspace.sortFoldersOnTop)) { child in
+					ProjectNavigatorItem(
+						item: child,
+						workspace: workspace,
+						windowController: windowController,
+						shouldloadChildren: $isExpanded,
+						selectedId: $selectedId
+					)
+				}
+			}
+		} label: {
+			HStack {
```

Comment:
No need to write again, just use `sidebarFileItem`?

Reply:
SidebarFileItem is too generic. Because we might have other "file items" in the sidebar as well

Reply:
All right. But `contextMenu` is missing on folder item.

Reply:
Fixed